File: impl/go/cli/cli_test.go
package cli

import (
	"bytes"
	"context"
	"io"
	"os"
	"path/filepath"
	"reflect"
	"strings"
	"sync"
	"testing"
	"time"

	twill "github.com/kt3k/twill/impl/go"
)

type syncBuffer struct {
	mu sync.Mutex
	b  bytes.Buffer
}

func (s *syncBuffer) Write(p []byte) (int, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.b.Write(p)
}

func (s *syncBuffer) String() string {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.b.String()
}

type result struct {
	code   int
	stdout string
	stderr string
}

func runCLI(t *testing.T, args []string, cwd string, stdin io.Reader) result {
	t.Helper()
	if cwd != "" {
		args = append([]string{"--cwd", cwd}, args...)
	}
	var stdout, stderr syncBuffer
	code := Main(context.Background(), args, stdin, &stdout, &stderr)
	return result{code, stdout.String(), stderr.String()}
}

func writeFile(t *testing.T, path, content string) {
	t.Helper()
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(path, []byte(content), 0o644); err != nil {
		t.Fatal(err)
	}
}

func project(t *testing.T) string {
	t.Helper()
	root := t.TempDir()
	writeFile(t, filepath.Join(root, "input.css"), "@import \"twill/theme.css\";\n@source not \"dist\";\n@twill utilities;\n")
	writeFile(t, filepath.Join(root, "index.html"), `<div class="flex p-4 hover:underline">`)
	writeFile(t, filepath.Join(root, "dist", "old.html"), `<div class="hidden">`)
	return root
}

func assertEqual(t *testing.T, got, want any) {
	t.Helper()
	if !reflect.DeepEqual(got, want) {
		t.Fatalf("got:\n%#v\nwant:\n%#v", got, want)
	}
}

func includes(t *testing.T, output, want string) {
	t.Helper()
	if !strings.Contains(output, want) {
		t.Fatalf("expected output to contain %q:\n%s", want, output)
	}
}

func TestParseArgs(t *testing.T) {
	defaults, err := ParseArgs(nil)
	if err != nil {
		t.Fatal(err)
	}
	assertEqual(t, defaults, &Options{Output: "-", Cwd: "."})

	full, err := ParseArgs([]string{"build", "-i", "a.css", "-o", "b.css", "--watch", "always", "--poll", "500", "-m", "--optimize", "--cwd", "x", "--silent"})
	if err != nil {
		t.Fatal(err)
	}
	assertEqual(t, full, &Options{Input: "a.css", Output: "b.css", Watch: WatchAlways, Poll: 500 * time.Millisecond, Minify: true, Optimize: true, Cwd: "x", Silent: true})

	watch, _ := ParseArgs([]string{"-w", "-i", "a.css"})
	assertEqual(t, watch.Watch, WatchOn)
	assertEqual(t, watch.Input, "a.css")
	always, _ := ParseArgs([]string{"--watch=always"})
	assertEqual(t, always.Watch, WatchAlways)
	poll, _ := ParseArgs([]string{"--poll"})
	assertEqual(t, poll.Poll, DefaultPollInterval)
	poll100, _ := ParseArgs([]string{"--poll=100"})
	assertEqual(t, poll100.Poll, 100*time.Millisecond)
	inline, _ := ParseArgs([]string{"--input=in.css", "--output=out.css"})
	assertEqual(t, inline.Input, "in.css")
	assertEqual(t, inline.Output, "out.css")
	help, _ := ParseArgs([]string{"-h"})
	assertEqual(t, help.Help, true)

	if _, err := ParseArgs([]string{"--poll=0"}); err == nil || !strings.Contains(err.Error(), "positive") {
		t.Fatalf("expected a positive interval error, got %v", err)
	}
	if _, err := ParseArgs([]string{"--nope"}); err == nil || !strings.Contains(err.Error(), "Unknown option") {
		t.Fatalf("expected an unknown option error, got %v", err)
	}
	if _, err := ParseArgs([]string{"-i"}); err == nil {
		t.Fatal("expected a missing value error")
	}
}

func TestAssembleSources(t *testing.T) {
	compiler := &twill.Compiler{Sources: []twill.SourceEntry{{Base: "/p", Pattern: "./src/**/*"}}}
	assertEqual(t, AssembleSources(compiler, "/p", "/p/in.css", "/bin/twill"), []twill.SourceEntry{
		{Base: "/p", Pattern: "**/*"},
		{Base: "/p", Pattern: "./src/**/*"},
		{Base: "/bin", Pattern: "twill", Negated: true},
		{Base: "/p", Pattern: "in.css"},
	})
	none := &twill.Compiler{Root: &twill.SourceRoot{None: true}}
	assertEqual(t, len(AssembleSources(none, "/p", "", "")), 0)
	rooted := &twill.Compiler{Root: &twill.SourceRoot{Base: "/p", Pattern: "../app"}}
	assertEqual(t, AssembleSources(rooted, "/p", "", ""), []twill.SourceEntry{{Base: "/p", Pattern: "../app"}})
}

func TestMinifyCSS(t *testing.T) {
	out, err := MinifyCSS(".a {\n  color: red !important;\n}\n@media (x) {\n  .b {\n    y: z;\n  }\n}\n")
	if err != nil {
		t.Fatal(err)
	}
	assertEqual(t, out, ".a{color:red!important;}@media (x){.b{y:z;}}")
}

func TestCLISingleBuild(t *testing.T) {
	root := project(t)
	r := runCLI(t, []string{"-i", "input.css"}, root, nil)
	assertEqual(t, r.code, 0)
	includes(t, r.stdout, ".flex {\n  display: flex;\n}")
	includes(t, r.stdout, ".p-4 {\n  padding: calc(var(--spacing) * 4);\n}")
	includes(t, r.stdout, ".hover\\:underline:hover")
	if strings.Contains(r.stdout, ".hidden") {
		t.Fatal("expected the excluded source to be skipped")
	}
	includes(t, r.stderr, "twill")
	includes(t, r.stderr, "Done in")

	toFile := runCLI(t, []string{"-i", "input.css", "-o", "out/build.css", "--silent"}, root, nil)
	assertEqual(t, toFile.code, 0)
	assertEqual(t, toFile.stdout, "")
	assertEqual(t, toFile.stderr, "")
	written, err := os.ReadFile(filepath.Join(root, "out", "build.css"))
	if err != nil {
		t.Fatal(err)
	}
	includes(t, string(written), ".flex {")

	minified := runCLI(t, []string{"-i", "input.css", "--minify", "--silent"}, root, nil)
	assertEqual(t, minified.code, 0)
	includes(t, minified.stdout, ".flex{display:flex;}")
}

func TestCLIStdinAndDefaultInput(t *testing.T) {
	root := project(t)
	r := runCLI(t, []string{"-i", "-", "--silent"}, root, strings.NewReader("@import \"twill/theme.css\";\n@twill utilities;"))
	assertEqual(t, r.code, 0)
	includes(t, r.stdout, ".flex {")

	defaulted := runCLI(t, []string{"--silent"}, root, strings.NewReader(""))
	assertEqual(t, defaulted.code, 0)
	includes(t, defaulted.stdout, "@layer theme, base, components, utilities;")
	includes(t, defaulted.stdout, ".flex {")
}

func TestCLIErrors(t *testing.T) {
	root := project(t)
	missing := runCLI(t, []string{"-i", "nope.css"}, root, nil)
	assertEqual(t, missing.code, 1)
	includes(t, missing.stderr, "does not exist")

	same := runCLI(t, []string{"-i", "input.css", "-o", "input.css"}, root, nil)
	assertEqual(t, same.code, 1)
	includes(t, same.stderr, "identical")

	writeFile(t, filepath.Join(root, "bad.css"), ".a { color: red;")
	bad := runCLI(t, []string{"-i", "bad.css", "--silent"}, root, nil)
	assertEqual(t, bad.code, 1)
	includes(t, bad.stderr, "Missing closing")

	unknown := runCLI(t, []string{"--nope"}, "", nil)
	assertEqual(t, unknown.code, 1)
	includes(t, unknown.stderr, "Unknown option")
	includes(t, unknown.stderr, "Usage: twill")
}

func TestCLIHelp(t *testing.T) {
	r := runCLI(t, []string{"--help"}, "", nil)
	assertEqual(t, r.code, 0)
	includes(t, r.stderr, "Usage: twill")
}

func waitFor(t *testing.T, check func() bool) {
	t.Helper()
	deadline := time.Now().Add(10 * time.Second)
	for time.Now().Before(deadline) {
		if check() {
			return
		}
		time.Sleep(50 * time.Millisecond)
	}
	t.Fatal("timed out waiting for the watcher")
}

func fileIncludes(path, text string) func() bool {
	return func() bool {
		content, err := os.ReadFile(path)
		return err == nil && strings.Contains(string(content), text)
	}
}

func TestCLIPollingRebuilds(t *testing.T) {
	root := project(t)
	output := filepath.Join(root, "out.css")
	ctx, cancel := context.WithCancel(context.Background())
	var stderr syncBuffer
	done := make(chan int)
	go func() {
		done <- Main(ctx, []string{"--cwd", root, "-i", "input.css", "-o", "out.css", "--poll", "50", "--silent"}, nil, io.Discard, &stderr)
	}()
	defer func() {
		cancel()
		<-done
	}()
	waitFor(t, fileIncludes(output, ".flex {"))
	// A source change is picked up incrementally.
	writeFile(t, filepath.Join(root, "page.html"), `<i class="hidden">`)
	waitFor(t, fileIncludes(output, ".hidden {"))
	// A stylesheet change triggers a full rebuild.
	writeFile(t, filepath.Join(root, "input.css"), "@import \"twill/theme.css\";\n@theme { --color-brand: blue; }\n@twill utilities;\n")
	writeFile(t, filepath.Join(root, "page.html"), `<i class="hidden bg-brand">`)
	waitFor(t, fileIncludes(output, ".bg-brand {"))
	if stderr.String() != "" {
		t.Fatalf("unexpected stderr: %s", stderr.String())
	}
}

func TestCLIWatchRebuildsAndExitsOnStdinEOF(t *testing.T) {
	root := project(t)
	output := filepath.Join(root, "out.css")
	stdinReader, stdinWriter := io.Pipe()
	var stderr syncBuffer
	done := make(chan int)
	go func() {
		done <- Main(context.Background(), []string{"--cwd", root, "-i", "input.css", "-o", "out.css", "--watch", "--silent"}, stdinReader, io.Discard, &stderr)
	}()
	waitFor(t, fileIncludes(output, ".flex {"))
	writeFile(t, filepath.Join(root, "page.html"), `<i class="hidden">`)
	waitFor(t, fileIncludes(output, ".hidden {"))
	writeFile(t, filepath.Join(root, "input.css"), "@import \"twill/theme.css\";\n@theme static { --color-brand: blue; }\n@twill utilities;\n")
	waitFor(t, fileIncludes(output, "--color-brand"))
	// A broken stylesheet reports an error and keeps watching.
	writeFile(t, filepath.Join(root, "input.css"), ".a { color: red;")
	waitFor(t, func() bool { return strings.Contains(stderr.String(), "Missing closing") })
	writeFile(t, filepath.Join(root, "input.css"), "@import \"twill/theme.css\";\n@theme static { --color-other: red; }\n@twill utilities;\n")
	waitFor(t, fileIncludes(output, "--color-other"))
	// A deleted source keeps the watcher alive.
	if err := os.Remove(filepath.Join(root, "page.html")); err != nil {
		t.Fatal(err)
	}
	writeFile(t, filepath.Join(root, "other.html"), `<i class="underline">`)
	waitFor(t, fileIncludes(output, ".underline {"))

	stdinWriter.Close()
	select {
	case code := <-done:
		assertEqual(t, code, 0)
	case <-time.After(5 * time.Second):
		t.Fatal("watch mode did not exit on stdin EOF")
	}
}

func TestCLIWatchAlwaysIgnoresStdinEOF(t *testing.T) {
	root := project(t)
	ctx, cancel := context.WithCancel(context.Background())
	done := make(chan int)
	go func() {
		done <- Main(ctx, []string{"--cwd", root, "-i", "input.css", "-o", "out.css", "--watch=always", "--silent"}, strings.NewReader(""), io.Discard, io.Discard)
	}()
	select {
	case <-done:
		t.Fatal("watch=always exited on stdin EOF")
	case <-time.After(500 * time.Millisecond):
	}
	cancel()
	assertEqual(t, <-done, 0)
}

File: impl/deno/cli_test.ts
import { assertEquals, assertStringIncludes, assertThrows } from "@std/assert";
import { join } from "@std/path";
import { assembleSources, minifyCss, parseCliArgs } from "./cli.ts";

const CLI = new URL("./cli.ts", import.meta.url).pathname;

async function runCli(
  args: string[],
  options: { cwd?: string; stdin?: string } = {},
): Promise<{ code: number; stdout: string; stderr: string }> {
  const command = new Deno.Command(Deno.execPath(), {
    args: ["run", "-A", CLI, ...args],
    cwd: options.cwd,
    stdin: options.stdin === undefined ? "null" : "piped",
    stdout: "piped",
    stderr: "piped",
  });
  const child = command.spawn();
  if (options.stdin !== undefined) {
    const writer = child.stdin.getWriter();
    await writer.write(new TextEncoder().encode(options.stdin));
    await writer.close();
  }
  const output = await child.output();
  return {
    code: output.code,
    stdout: new TextDecoder().decode(output.stdout),
    stderr: new TextDecoder().decode(output.stderr),
  };
}

async function project(): Promise<string> {
  const root = await Deno.makeTempDir({ prefix: "twill-cli-" });
  await Deno.writeTextFile(
    join(root, "input.css"),
    `@import "twill/theme.css";\n@source not "dist";\n@twill utilities;\n`,
  );
  await Deno.writeTextFile(
    join(root, "index.html"),
    `<div class="flex p-4 hover:underline">`,
  );
  await Deno.mkdir(join(root, "dist"));
  await Deno.writeTextFile(
    join(root, "dist", "old.html"),
    `<div class="hidden">`,
  );
  return root;
}

Deno.test("parseCliArgs", () => {
  assertEquals(parseCliArgs([]), {
    input: null,
    output: "-",
    watch: false,
    poll: null,
    minify: false,
    optimize: false,
    cwd: ".",
    silent: false,
    help: false,
  });
  const full = parseCliArgs([
    "build",
    "-i",
    "a.css",
    "-o",
    "b.css",
    "--watch",
    "always",
    "--poll",
    "500",
    "-m",
    "--optimize",
    "--cwd",
    "x",
    "--silent",
  ]);
  assertEquals(full, {
    input: "a.css",
    output: "b.css",
    watch: "always",
    poll: 500,
    minify: true,
    optimize: true,
    cwd: "x",
    silent: true,
    help: false,
  });
  assertEquals(parseCliArgs(["-w", "-i", "a.css"]).watch, true);
  assertEquals(parseCliArgs(["--watch=always"]).watch, "always");
  assertEquals(parseCliArgs(["--poll"]).poll, 250);
  assertEquals(parseCliArgs(["--poll=100"]).poll, 100);
  assertEquals(parseCliArgs(["-h"]).help, true);
  assertThrows(() => parseCliArgs(["--poll=0"]), Error, "positive");
  assertThrows(() => parseCliArgs(["--nope"]), Error, "Unknown option");
});

Deno.test("assembleSources", () => {
  const compiler = {
    sources: [{ base: "/p", pattern: "./src/**/*", negated: false }],
    root: null,
    features: 0,
    designSystem: null as unknown as never,
    build: () => "",
  };
  assertEquals(
    assembleSources(compiler, {
      cwd: "/p",
      inputPath: "/p/in.css",
      executable: "/bin/twill",
    }),
    [
      { base: "/p", pattern: "**/*", negated: false },
      { base: "/p", pattern: "./src/**/*", negated: false },
      { base: "/bin", pattern: "twill", negated: true },
      { base: "/p", pattern: "in.css", negated: false },
    ],
  );
  assertEquals(
    assembleSources({ ...compiler, root: "none", sources: [] }, {
      cwd: "/p",
      inputPath: null,
      executable: null,
    }),
    [],
  );
  assertEquals(
    assembleSources({
      ...compiler,
      root: { base: "/p", pattern: "../app" },
      sources: [],
    }, {
      cwd: "/p",
      inputPath: null,
      executable: null,
    }),
    [{ base: "/p", pattern: "../app", negated: false }],
  );
});

Deno.test("minifyCss", () => {
  assertEquals(
    minifyCss(
      ".a {\n  color: red !important;\n}\n@media (x) {\n  .b {\n    y: z;\n  }\n}\n",
    ),
    ".a{color:red!important;}@media (x){.b{y:z;}}",
  );
});

Deno.test("cli: single build to stdout and to a file", async () => {
  const root = await project();
  try {
    const result = await runCli(["-i", "input.css"], { cwd: root });
    assertEquals(result.code, 0, result.stderr);
    assertStringIncludes(result.stdout, ".flex {\n  display: flex;\n}");
    assertStringIncludes(
      result.stdout,
      ".p-4 {\n  padding: calc(var(--spacing) * 4);\n}",
    );
    assertStringIncludes(result.stdout, ".hover\\:underline:hover");
    assertEquals(result.stdout.includes(".hidden"), false);
    assertStringIncludes(result.stderr, "twill");
    assertStringIncludes(result.stderr, "Done in");

    const toFile = await runCli([
      "-i",
      "input.css",
      "-o",
      "out/build.css",
      "--silent",
    ], { cwd: root });
    assertEquals(toFile.code, 0, toFile.stderr);
    assertEquals(toFile.stdout, "");
    assertEquals(toFile.stderr, "");
    const written = await Deno.readTextFile(join(root, "out", "build.css"));
    assertStringIncludes(written, ".flex {");

    const minified = await runCli(["-i", "input.css", "--minify", "--silent"], {
      cwd: root,
    });
    assertEquals(minified.code, 0, minified.stderr);
    assertStringIncludes(minified.stdout, ".flex{display:flex;}");
  } finally {
    await Deno.remove(root, { recursive: true });
  }
});

Deno.test("cli: stdin input and default input", async () => {
  const root = await project();
  try {
    const result = await runCli(["-i", "-", "--silent"], {
      cwd: root,
      stdin: `@import "twill/theme.css";\n@twill utilities;`,
    });
    assertEquals(result.code, 0, result.stderr);
    assertStringIncludes(result.stdout, ".flex {");

    const defaulted = await runCli(["--silent"], { cwd: root, stdin: "" });
    assertEquals(defaulted.code, 0, defaulted.stderr);
    assertStringIncludes(
      defaulted.stdout,
      "@layer theme, base, components, utilities;",
    );
    assertStringIncludes(defaulted.stdout, ".flex {");
  } finally {
    await Deno.remove(root, { recursive: true });
  }
});

Deno.test("cli: errors exit with status 1", async () => {
  const root = await project();
  try {
    const missing = await runCli(["-i", "nope.css"], { cwd: root });
    assertEquals(missing.code, 1);
    assertStringIncludes(missing.stderr, "does not exist");

    const same = await runCli(["-i", "input.css", "-o", "input.css"], {
      cwd: root,
    });
    assertEquals(same.code, 1);
    assertStringIncludes(same.stderr, "identical");

    await Deno.writeTextFile(join(root, "bad.css"), ".a { color: red;");
    const bad = await runCli(["-i", "bad.css", "--silent"], { cwd: root });
    assertEquals(bad.code, 1);
    assertStringIncludes(bad.stderr, "Missing closing");
  } finally {
    await Deno.remove(root, { recursive: true });
  }
});

Deno.test("cli: --help", async () => {
  const result = await runCli(["--help"]);
  assertEquals(result.code, 0);
  assertStringIncludes(result.stderr, "Usage: twill");
});

async function waitFor(
  check: () => Promise<boolean>,
  timeout = 10000,
): Promise<void> {
  const start = Date.now();
  while (Date.now() - start < timeout) {
    if (await check()) return;
    await new Promise((r) => setTimeout(r, 100));
  }
  throw new Error("Timed out waiting for the watcher");
}

async function fileIncludes(path: string, text: string): Promise<boolean> {
  try {
    return (await Deno.readTextFile(path)).includes(text);
  } catch {
    return false;
  }
}

Deno.test("cli: polling mode rebuilds on changes", async () => {
  const root = await project();
  const output = join(root, "out.css");
  const command = new Deno.Command(Deno.execPath(), {
    args: [
      "run",
      "-A",
      CLI,
      "-i",
      "input.css",
      "-o",
      "out.css",
      "--poll",
      "100",
      "--silent",
    ],
    cwd: root,
    stdin: "null",
    stdout: "null",
    stderr: "piped",
  });
  const child = command.spawn();
  try {
    await waitFor(() => fileIncludes(output, ".flex {"));
    // A source change is picked up incrementally.
    await Deno.writeTextFile(join(root, "page.html"), `<i class="hidden">`);
    await waitFor(() => fileIncludes(output, ".hidden {"));
    // A stylesheet change triggers a full rebuild.
    await Deno.writeTextFile(
      join(root, "input.css"),
      `@import "twill/theme.css";\n@theme { --color-brand: blue; }\n@twill utilities;\n`,
    );
    await Deno.writeTextFile(
      join(root, "page.html"),
      `<i class="hidden bg-brand">`,
    );
    await waitFor(() => fileIncludes(output, ".bg-brand {"));
  } finally {
    child.kill("SIGTERM");
    await child.status;
    await Deno.remove(root, { recursive: true });
  }
});

Deno.test("cli: watch mode rebuilds on changes and exits on stdin EOF", async () => {
  const root = await project();
  const output = join(root, "out.css");
  const command = new Deno.Command(Deno.execPath(), {
    args: [
      "run",
      "-A",
      CLI,
      "-i",
      "input.css",
      "-o",
      "out.css",
      "--watch",
      "--silent",
    ],
    cwd: root,
    stdin: "piped",
    stdout: "null",
    stderr: "piped",
  });
  const child = command.spawn();
  let stderr = "";
  const reading = (async () => {
    for await (const chunk of child.stderr) {
      stderr += new TextDecoder().decode(chunk);
    }
  })();
  try {
    await waitFor(() => fileIncludes(output, ".flex {"));
    await Deno.writeTextFile(join(root, "page.html"), `<i class="hidden">`);
    await waitFor(() => fileIncludes(output, ".hidden {"));
    await Deno.writeTextFile(
      join(root, "input.css"),
      `@import "twill/theme.css";\n@theme static { --color-brand: blue; }\n@twill utilities;\n`,
    );
    await waitFor(() => fileIncludes(output, "--color-brand"));
    // A broken stylesheet reports an error and keeps watching. The file is
    // written twice so that a rebuild sees the broken content even if the
    // first event fired between truncation and write.
    await Deno.writeTextFile(join(root, "input.css"), `.a { color: red;`);
    await new Promise((r) => setTimeout(r, 300));
    await Deno.writeTextFile(join(root, "input.css"), `.a { color: red;`);
    await waitFor(() => Promise.resolve(stderr.includes("Missing closing")));
    await Deno.writeTextFile(
      join(root, "input.css"),
      `@import "twill/theme.css";\n@theme static { --color-other: red; }\n@twill utilities;\n`,
    );
    await waitFor(() => fileIncludes(output, "--color-other"));
  } finally {
    await child.stdin.close();
    const status = await Promise.race([
      child.status,
      new Promise<null>((r) => setTimeout(() => r(null), 5000)),
    ]);
    if (status === null) child.kill("SIGTERM");
    await child.status;
    await reading;
    await Deno.remove(root, { recursive: true });
  }
});
